### PR TITLE
AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

### PR DESCRIPTION
**Pull request by changpeng**
This work simplifies and generalizes the instruction definition for intrinsic llvm.fptrunc.round. We no longer name the instruction with the rounding mode. Instead, we introduce an immediate operand for the rounding mode for the pseudo instruction. This immediate will be used to set up the hardware mode register at the time the real instruction is generated. We name the pseudo instruction as FPTRUNC_ROUND_F16_F32 (for f32 -> f16), which is easy to generalize for other types.

  "round.towardzero" and "round.tonearest" are added for f32 -> f16 truncating, in addition to the existing "round.upward" and "round.downward". Other rounding modes are not supported by hardware at this moment.

### CODE REVIEW

**Comment by jayfoad** [comment]
> Improve codegen for intrinsic llvm.fptrunc.round

Do you have an example where codegen is improved?

> We name the pseudo instruction as FPTRUNC_ROUND_F32_TO_F16 (for f32 -> f16)

How about calling it `FPTRUNC_ROUND_F16_F32` to match `V_CVT_F16_F32`?

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll [hunk 1]
```diff
@@ -3,6 +3,15 @@
 ; RUN: llc -global-isel=0 -mtriple=amdgcn -mcpu=gfx1010 < %s | FileCheck -check-prefixes=CHECK,SDAG %s
 ; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx1030 < %s | FileCheck -check-prefixes=CHECK,GISEL %s
 
+define amdgpu_gs half @v_fptrunc_round_f32_to_f16_tonearest(float %a) {
+; CHECK-LABEL: v_fptrunc_round_f32_to_f16_tonearest:
+; CHECK:       ; %bb.0:
+; CHECK-NEXT:    v_cvt_f16_f32_e32 v0, v0
+; CHECK-NEXT:    ; return to shader part epilog
+  %res = call half @llvm.fptrunc.round.f16.f32(float %a, metadata !"round.tonearest")
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: this should work, but hopefully instcombine just turns this into fptrunc?

**Review by arsenm — COMMENTED** [review]
Title is misleading, this doesn't change codegen at all?

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll (hunk 1 above)
Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: At custom lowering the the intrinsic, we can turn it to fp_round, which can also generated the expected output for "round.tonearest". However, it seems the lowering of fp_round ignores the rounding mode operand in most cases (just assume default of round-to-nearest-even)

**Comment by changpeng** [comment]
> Title is misleading, this doesn't change codegen at all?
The patch changes the instruction definition, and adds the cases for two additional rounding modes. I was thinking, the whole process for intrinsic to generate assembly code is called code generation. But codegen may have a specific meaning.

**Comment by changpeng** [comment]
> Do you have an example where codegen is improved?
Same code generated for round.upward ad round.downward. Add two additional cases round.tonearst and round.towardzero.
 
> How about calling it `FPTRUNC_ROUND_F16_F32` to match `V_CVT_F16_F32`?
Will do as suggested. Thanks.

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: that shouldn't be custom lowering, that should be a generic lowering (and instcombine optimization)

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 2]
```diff
@@ -7089,23 +7089,29 @@ bool AMDGPULegalizerInfo::legalizeBVHIntrinsic(MachineInstr &MI,
...
 
-  if (RoundMode == (int)RoundingMode::TowardPositive)
-    Opc = AMDGPU::G_FPTRUNC_ROUND_UPWARD;
-  else if (RoundMode == (int)RoundingMode::TowardNegative)
-    Opc = AMDGPU::G_FPTRUNC_ROUND_DOWNWARD;
-  else
+  // Only support towardzero, tonearest, upward and downward.
+  int RoundMode = MI.getOperand(2).getImm();
+  if (RoundMode < (int)RoundingMode::TowardZero ||
+      RoundMode > (int)RoundingMode::TowardNegative)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:7101]: Do exact equality checks for the supported modes

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 3]
```diff
@@ -6671,10 +6655,34 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
...
 
+SDValue SITargetLowering::lowerFPTRUNC_ROUND(SDValue Op,
+                                             SelectionDAG &DAG) const {
+  if (Op.getOperand(0)->getValueType(0) != MVT::f32)
+    return SDValue();
+
+  // Only support towardzero, tonearest, upward and downward.
+  int RoundMode = Op.getConstantOperandVal(1);
+  if (RoundMode < (int)RoundingMode::TowardZero ||
+      RoundMode > (int)RoundingMode::TowardNegative)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:6666]: Same, do exact equality checks for the supported modes

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 4]
```diff
@@ -6671,10 +6655,34 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
...
+  unsigned Opc = AMDGPUISD::FPTRUNC_ROUND;
+  return DAG.getNode(Opc, DL, Op.getNode()->getVTList(), Op->getOperand(0),
+                     DAG.getTargetConstant(HW_Mode, DL, MVT::i32));
+}
+
 SDValue SITargetLowering::lowerFP_ROUND(SDValue Op, SelectionDAG &DAG) const {
   assert(Op.getValueType() == MVT::f16 &&
          "Do not know how to custom lower FP_ROUND for non-f16 type");
 
+  return Op;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:6684]: Making the rest of the function dead, which must break something

Comment (changpeng) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:6684]: Sorry, this was the leftover when I tried something else. should not change fp_round

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll (hunk 1 above)
Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: I would put the instcombine optimization for llvm.fptrunc.round of mode round.tonearest to a separate patch. Thanks.

Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: Can we do something in instcombine for round.towardzero also?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: > I would put the instcombine optimization for llvm.fptrunc

Yes

> Can we do something in instcombine for round.towardzero also?

I assume quite a few combines could be done with it, but you can't merely replace it with anything obvious

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 5]
```diff
@@ -2180,7 +2180,7 @@ bool AMDGPULegalizerInfo::legalizeCustom(
   case TargetOpcode::G_CTLZ_ZERO_UNDEF:
     return legalizeCTLZ_ZERO_UNDEF(MI, MRI, B);
   case TargetOpcode::G_INTRINSIC_FPTRUNC_ROUND:
-    return legalizeFPTruncRound(MI, B);
+    return legalizeFPTruncRound(MI, MRI, B);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:2183]: Don't need to add this parameter, you can just get MRI from the MachineIRBuilder

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 6]
```diff
@@ -7089,23 +7089,31 @@ bool AMDGPULegalizerInfo::legalizeBVHIntrinsic(MachineInstr &MI,
...
+  // "round.towardzero" -> TowardZero 0        -> FP_ROUND_ROUND_TO_ZERO 3
+  // "round.tonearest"  -> NearestTiesToEven 1 -> FP_ROUND_ROUND_TO_NEAREST 0
+  // "round.upward"     -> TowardPositive 2    -> FP_ROUND_ROUND_TO_INF 1
+  // "round.downward    -> TowardNegative 3    -> FP_ROUND_ROUND_TO_NEGINF 2
+  unsigned HW_Mode = (RoundMode + 3) % 4;
+  B.buildInstr(AMDGPU::G_FPTRUNC_ROUND)
       .addDef(MI.getOperand(0).getReg())
-      .addUse(MI.getOperand(1).getReg());
+      .addUse(Src)
+      .addImm(HW_Mode);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:7114]: All this is really doing is just remapping the encoded rounding mode values. At that point it should just be a selection pattern

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 7]
```diff
@@ -225,20 +225,12 @@ def S_INVERSE_BALLOT_U64 : SPseudoInstSI<
...
+// in the ModeRegister pass.
 let Uses = [MODE, EXEC] in {
-def FPTRUNC_UPWARD_PSEUDO : VPseudoInstSI <(outs VGPR_32:$vdst),
-  (ins VGPR_32:$src0),
-  [(set f16:$vdst, (SIfptrunc_round_upward f32:$src0))]>;
-
-def FPTRUNC_DOWNWARD_PSEUDO : VPseudoInstSI <(outs VGPR_32:$vdst),
-  (ins VGPR_32:$src0),
-  [(set f16:$vdst, (SIfptrunc_round_downward f32:$src0))]>;
+def FPTRUNC_ROUND_F16_F32 : VPseudoInstSI <(outs VGPR_32:$vdst),
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:231]: This needs to be more obviously a pseudo from the name

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll [hunk 8]
```diff
@@ -23,6 +32,16 @@ define amdgpu_gs half @v_fptrunc_round_f32_to_f16_downward(float %a) {
...
+define amdgpu_gs half @v_fptrunc_round_f32_to_f16_towardzero(float %a) {
+; CHECK-LABEL: v_fptrunc_round_f32_to_f16_towardzero:
+; CHECK:       ; %bb.0:
+; CHECK-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 3
+; CHECK-NEXT:    v_cvt_f16_f32_e32 v0, v0
+; CHECK-NEXT:    ; return to shader part epilog
+  %res = call half @llvm.fptrunc.round.f16.f32(float %a, metadata !"round.towardzero")
+  ret half %res
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:44]: can you also include a test where multiple modes are used in the same function

Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:44]: Add tests that has multiple calls with round.towardzero and round.downward.

File: llvm/lib/Target/AMDGPU/SIInstructions.td (hunk 7 above)
Comment (changpeng) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:231]: FPTRUNC_ROUND_PSEUDO_F16_F32 ?

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll (hunk 1 above)
Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: For f32 -> bf16, you won't be able to rtz for v_cvt_pk_bf16_f32 (hardware uses round to nearest even all the times).
So if you really want to rtz, you will have to do in the compiler.

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 6 above)
Comment (changpeng) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:7114]: Do you mean we can use a patch to directly map intrinsic to pseudo? I can try that in a separate patch.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:7114]: Yes

File: llvm/lib/Target/AMDGPU/SIInstructions.td (hunk 7 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:231]: FPTRUNC_ROUND_F16_F32_PSEUDO?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 9]
```diff
@@ -6671,6 +6655,30 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
...
+      RoundMode != (int)RoundingMode::TowardNegative)
+    return SDValue();
+
+  // "round.towardzero" -> TowardZero 0        -> FP_ROUND_ROUND_TO_ZERO 3
+  // "round.tonearest"  -> NearestTiesToEven 1 -> FP_ROUND_ROUND_TO_NEAREST 0
+  // "round.upward"     -> TowardPositive 2    -> FP_ROUND_ROUND_TO_INF 1
+  // "round.downward    -> TowardNegative 3    -> FP_ROUND_ROUND_TO_NEGINF 2
+  unsigned HW_Mode = (RoundMode + 3) % 4;
+  SDLoc DL(Op);
+  unsigned Opc = AMDGPUISD::FPTRUNC_ROUND;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:6677]: can lose the Opc variable (or just move this all to a pattern)

File: llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:11]: Yes, it will need some kind of expansion

**Review by arsenm — APPROVED** [review]
lgtm but this should still just be a pattern

**Comment by changpeng** [comment]
> lgtm but this should still just be a pattern

Do you mean the pattern to map "ISD::FPTRUNC_ROUND src, round" to pseudo?
I am not sure how to translate round to the hardware rounding mode during the pattern matching!

**Comment by arsenm** [comment]
> > lgtm but this should still just be a pattern
> 
> Do you mean the pattern to map "ISD::FPTRUNC_ROUND src, round" to pseudo? I am not sure how to translate round to the hardware rounding mode during the pattern matching!

Yes. You just need an SDNodeXForm to modify the value

**Comment by changpeng** [comment]
> Yes. You just need an SDNodeXForm to modify the value

What about a rounding mode that is not supported, such as "round.tonearestaway"? 

I am doing like following, but it does not show that the intrinsic can not be selected:

def as_supported_rounding_mode: SDNodeXForm<imm, [{
  int RoundMode = N->getSExtValue();
  if (RoundMode == (int)RoundingMode::TowardZero ||
      RoundMode == (int)RoundingMode::NearestTiesToEven ||
      RoundMode == (int)RoundingMode::TowardPositive ||
      RoundMode == (int)RoundingMode::TowardNegative) {
    unsigned HW_Mode = (RoundMode + 3) % 4;
    return CurDAG->getTargetConstant(HW_Mode, SDLoc(N), MVT::i32);
  }
  return SDValue();
}]>;

**Comment by changpeng** [comment]
And what to do on global-isel path? I guess something similar. 
It will end up with similar complexity as before.

**Comment by arsenm** [comment]
> And what to do on global-isel path? I guess something similar. It will end up with similar complexity as before.

You use the same pattern, the SDNodeXForm implementation is just different and uses a custom renderer function. The overall thing will be 1/5th the code you have here for the 2x custom lowering

**Comment by arsenm** [comment]
> > Yes. You just need an SDNodeXForm to modify the value
> 
> What about a rounding mode that is not supported, such as "round.tonearestaway"?
> 

You can use a TImmLeaf predicate on the immediate value (which can also combine the xform)

**Pull request merged**